processing/src/main/java/org/apache/druid/query/context/ResponseContext.java

```diff
@@ -236,95 +253,124 @@ public static ResponseContext createEmpty()
-   * Merges a response context into current.
-   * This method merges only keys from the enum {@link Key}.
+   * Merges a response context into the current.
+   * @throws IllegalStateException If a key of the {@code responseContext} has not been registered.
    */
   public void merge(ResponseContext responseContext)
   {
-    for (BaseKey key : Key.getKeys()) {
-      final Object newValue = responseContext.get(key);
+    responseContext.getDelegate().forEach((key, newValue) -> {
       if (newValue != null) {
         add(key, newValue);
       }
-    }
+    });
   }
 
   /**
    * Serializes the context given that the resulting string length is less than the provided limit.
-   * The method removes max-length fields one by one if the resulting string length is greater than the limit.
-   * The resulting string might be correctly deserialized as a {@link ResponseContext}.
+   * This method tries to remove some elements from context collections if it's needed to satisfy the limit.
+   * The resulting string might be correctly deserialized to {@link ResponseContext}.
    */
   public SerializationResult serializeWith(ObjectMapper objectMapper, int maxCharsNumber) throws JsonProcessingException
   {
     final String fullSerializedString = objectMapper.writeValueAsString(getDelegate());
     if (fullSerializedString.length() <= maxCharsNumber) {
       return new SerializationResult(fullSerializedString, fullSerializedString);
     } else {
-      final HashMap<String, Object> copiedMap = new HashMap<>(getDelegate());
-      final PriorityQueue<Map.Entry<String, String>> serializedValueEntries = new PriorityQueue<>(
-          Comparator.comparing((Map.Entry<String, String> e) -> e.getValue().length()).reversed()
-      );
-      for (Map.Entry<String, Object> e : copiedMap.entrySet()) {
-        serializedValueEntries.add(new AbstractMap.SimpleImmutableEntry<>(
-            e.getKey(),
-            objectMapper.writeValueAsString(e.getValue())
-        ));
-      }
-      // quadratic complexity: while loop with map serialization on each iteration
-      while (!copiedMap.isEmpty() && !serializedValueEntries.isEmpty()) {
-        final Map.Entry<String, String> maxLengthEntry = serializedValueEntries.poll();
-        Preconditions.checkNotNull(maxLengthEntry);
-        copiedMap.remove(maxLengthEntry.getKey());
-        final String reducedSerializedString = objectMapper.writeValueAsString(copiedMap);
-        if (reducedSerializedString.length() <= maxCharsNumber) {
-          return new SerializationResult(reducedSerializedString, fullSerializedString);
+      // Indicates that the context is truncated during serialization.
+      add(Key.TRUNCATED, true);
+      final ObjectNode contextJsonNode = objectMapper.valueToTree(getDelegate());
+      final ArrayList<Map.Entry<String, JsonNode>> sortedNodesByLength = Lists.newArrayList(contextJsonNode.fields());
+      final Comparator<Map.Entry<String, JsonNode>> valueLengthReversedComparator =
```

Comment:
Please extract this comparator as a constant.

Reply:
Extracted